File: frontend/src/api/accounting/synthese.ts
import type { Sens } from './common';

/** Consolidated dashboard payload (T6). All amounts are decimal strings. */
export interface SyntheseResultat {
  recettes: string;
  depenses: string;
  resultat: string;
}

export interface RepartitionCategorie {
  categorie_id: string;
  libelle: string;
  sens: Sens;
  montant: string;
}

export interface RepartitionEvenement {
  evenement_id: string;
  nom: string;
  couleur: string | null;
  recettes: string;
  depenses: string;
  resultat: string;
}

export interface RepartitionTresorerie {
  compte_id: string;
  libelle: string;
  recettes: string;
  depenses: string;
}

export interface CourbePoint {
  date: string;
  solde: string;
}

export interface AlerteEvenement {
  evenement_id: string;
  nom: string;
  budget_depenses: string;
  realise_depenses: string;
}

export interface AlerteExercice {
  exercice_id: string;
  libelle: string;
  date_fin: string;
}

export interface AlerteBudget {
  categorie_id: string;
  libelle: string;
  montant_prevu: string;
  realise: string;
}

export interface SyntheseAlertes {
  brouillons: number;
  evenements_depasses: AlerteEvenement[];
  exercices_a_cloturer: AlerteExercice[];
  budgets_depasses: AlerteBudget[];
}

/** Dashboard budget widget: prévu vs réalisé of the period's exercice budget. */
export interface BudgetSynthese {
  exercice_id: string;
  exercice_libelle: string;
  recettes_prevu: string;
  recettes_realise: string;
  depenses_prevu: string;
  depenses_realise: string;
  resultat_prevu: string;
  resultat_realise: string;
  depassements: AlerteBudget[];
}

export interface Synthese {
  date_from: string;
  date_to: string;
  resultat: SyntheseResultat;
  repartition_categories: RepartitionCategorie[];
  repartition_evenements: RepartitionEvenement[];
  repartition_tresorerie: RepartitionTresorerie[];
  courbe_tresorerie: CourbePoint[];
  alertes: SyntheseAlertes;
  budget: BudgetSynthese | null;
}

/** Period for the synthesis; omit both to let the server use the open exercice. */
export interface SyntheseParams {
  date_from?: string;
  date_to?: string;
}

File: backend/models/synthese.py
from datetime import date
from decimal import Decimal

from sqlmodel import SQLModel

from .categorie import SensCategorie


class SyntheseResultat(SQLModel):
    """Result over the period: produits (cl.7) − charges (cl.6)."""

    recettes: Decimal
    depenses: Decimal
    resultat: Decimal


class RepartitionCategorieItem(SQLModel):
    """One slice of the per-category breakdown over the period."""

    categorie_id: str
    libelle: str
    sens: SensCategorie
    montant: Decimal


class RepartitionEvenementItem(SQLModel):
    """One slice of the per-event breakdown over the period."""

    evenement_id: str
    nom: str
    couleur: str | None
    recettes: Decimal
    depenses: Decimal
    resultat: Decimal


class CourbePoint(SQLModel):
    """One point of the treasury balance curve (cumulative end-of-day balance)."""

    date: date
    solde: Decimal


class AlerteEvenement(SQLModel):
    evenement_id: str
    nom: str
    budget_depenses: Decimal
    realise_depenses: Decimal


class AlerteExercice(SQLModel):
    exercice_id: str
    libelle: str
    date_fin: date


class AlerteBudget(SQLModel):
    """A budgeted dépense category whose réalisé exceeds its prévu amount."""

    categorie_id: str
    libelle: str
    montant_prevu: Decimal
    realise: Decimal


class SyntheseAlertes(SQLModel):
    """Current actionable alerts (independent of the selected period)."""

    brouillons: int  # entries still in draft, to validate
    evenements_depasses: list[AlerteEvenement]
    exercices_a_cloturer: list[AlerteExercice]
    budgets_depasses: list[AlerteBudget]


class BudgetSynthese(SQLModel):
    """Dashboard budget widget: prévu vs réalisé of the period's exercice budget."""

    exercice_id: str
    exercice_libelle: str
    recettes_prevu: Decimal
    recettes_realise: Decimal
    depenses_prevu: Decimal
    depenses_realise: Decimal
    resultat_prevu: Decimal
    resultat_realise: Decimal
    depassements: list[AlerteBudget]


class SyntheseRead(SQLModel):
    """Consolidated dashboard: period analytics + current alerts, in one read."""

    date_from: date
    date_to: date
    resultat: SyntheseResultat
    repartition_categories: list[RepartitionCategorieItem]
    repartition_evenements: list[RepartitionEvenementItem]
    courbe_tresorerie: list[CourbePoint]
    alertes: SyntheseAlertes
    budget: "BudgetSynthese | None" = None
